File: bindings/python/examples/send_modes.py
"""Sweeps a focus around a circle two ways: stop-and-wait vs streaming.

Both languages pipeline the streaming mode with two-stage await: `send` enqueues a
frame and returns a ResponseFuture, and a FIFO deque bounds the in-flight frames.
Run with: cargo xtask py example send_modes"""

import asyncio
import collections
import math
import time

import numpy as np

import autd3
import autd3_link_ethercrab as ethercrab
import autd3_pattern as pattern
from autd3.units import m, s

TOTAL_POINTS = 1000


def report(label: str, elapsed: float) -> None:
    rate = TOTAL_POINTS / elapsed
    print(f"{label}: {TOTAL_POINTS} updates in {elapsed:.2f}s ({rate:.0f} updates/s)")


async def configure(client: autd3.Client, patterns: object) -> None:
    pattern.null(patterns)
    builder = client.datagram_builder()
    builder.push(autd3.commands.WritePatternBuffer(autd3.value.PatternBank.B0, 0, patterns))
    builder.push(autd3.commands.ConfigPattern(autd3.value.PatternBank.B0, autd3.value.SamplingConfig.FREQ_4K, 1))
    for frame in builder.build():
        await client.send_checked(frame)


def write_focus(client: autd3.Client, patterns: object) -> object:
    builder = client.datagram_builder()
    builder.push(autd3.commands.WritePatternBuffer(autd3.value.PatternBank.B0, 0, patterns))
    return builder.build()


async def main() -> None:
    geometry = autd3.geometry.Geometry([autd3.geometry.Autd3([0.0, 0.0, 0.0], [1.0, 0.0, 0.0, 0.0])])

    client = await autd3.Client.open(
        geometry,
        ethercrab.EtherCrabLinkOption(),
        autd3.ClientConfig(),
    )

    center = geometry.center()
    radius = 30.0
    wavelength = pattern.wavelength(340 * m / s)

    patterns = geometry.pattern_buffer()
    await configure(client, patterns)

    datagrams = []
    for i in range(TOTAL_POINTS):
        theta = 2.0 * math.pi * i / TOTAL_POINTS
        target = center + np.array([radius * math.cos(theta), radius * math.sin(theta), 150.0])
        pattern.focus(geometry, target, wavelength, pattern.FocusOption(), patterns)
        datagrams.append(write_focus(client, patterns))

    print(f"sweeping a focus through {TOTAL_POINTS} positions, twice")

    # stop-and-wait: confirm each frame lands before issuing the next.
    start = time.perf_counter()
    for dg in datagrams:
        for frame in dg:
            await client.send_checked(frame)
    report("stop-and-wait", time.perf_counter() - start)

    # streaming: keep MAX_IN_FLIGHT frames on the wire, draining the oldest response
    # once the window is full.
    start = time.perf_counter()
    pending = collections.deque()
    for dg in datagrams:
        for frame in dg:
            if len(pending) >= autd3.MAX_IN_FLIGHT:
                await pending.popleft()
            pending.append(await client.send(frame))
    while pending:
        await pending.popleft()
    report("streaming", time.perf_counter() - start)

    await client.stop()
    await client.close()


if __name__ == "__main__":
    asyncio.run(main())

File: examples/src/bin/send_modes.rs
// Sweeps a focus around a circle, sending the same update sequence two ways to
// show how the send loop chooses its mode. Run with: cargo xtask example send_modes

use std::collections::VecDeque;
use std::f32::consts::PI;
use std::time::{Duration, Instant};

use anyhow::Result;

use autd3_rs::geometry::{Autd3, Geometry, Point3, offset};
use autd3_rs::params::NUM_TRANSDUCERS;
use autd3_rs::units::{m, mm, s};
use autd3_rs::value::{Emission, LoopBehavior, PatternBank, SamplingConfig};
use autd3_rs::{
    Client, ClientConfig, ConfigPattern, Frames, Length, MAX_IN_FLIGHT, ResponseFuture,
    WritePatternBuffer,
};
use autd3_rs_link_ethercrab::EtherCrabLinkOption;

const TOTAL_POINTS: usize = 1000;

#[tokio::main(flavor = "multi_thread")]
async fn main() -> Result<()> {
    tracing_subscriber::fmt()
        .with_env_filter(
            tracing_subscriber::EnvFilter::try_from_default_env()
                .unwrap_or_else(|_| tracing_subscriber::EnvFilter::new("info")),
        )
        .init();

    let geometry = Geometry::new(vec![Autd3::default()]);

    let client = Client::open(
        &geometry,
        EtherCrabLinkOption::default(),
        ClientConfig::default(),
    )
    .await?;

    configure(&client, &geometry).await?;

    let center = geometry.center();
    let radius = 30.0 * mm;
    let wavelength = autd3_rs_pattern::wavelength(340.0 * m / s);
    let targets: Vec<Point3<f32>> = (0..TOTAL_POINTS)
        .map(|i| {
            let theta = 2. * PI * (i as f32) / (TOTAL_POINTS as f32);
            center + offset(radius * theta.cos(), radius * theta.sin(), 150.0 * mm)
        })
        .collect();

    println!("sweeping a focus through {TOTAL_POINTS} positions, twice");

    let elapsed = run_stop_and_wait(&client, &geometry, &targets, wavelength).await?;
    report("stop-and-wait", elapsed);

    let elapsed = run_streaming(&client, &geometry, &targets, wavelength, MAX_IN_FLIGHT).await?;
    report("streaming", elapsed);

    client.stop().await?;
    client.close().await?;

    Ok(())
}

// One round-trip per frame: confirm each update lands before issuing the next.
async fn run_stop_and_wait(
    client: &Client,
    geometry: &Geometry,
    targets: &[Point3<f32>],
    wavelength: Length,
) -> Result<Duration> {
    let mut emissions = geometry.pattern_buffer();
    let mut buf = Frames::default();

    let start = Instant::now();
    for &target in targets {
        autd3_rs_pattern::focus(
            geometry,
            target,
            wavelength,
            &autd3_rs_pattern::FocusOption::default(),
            &mut emissions,
        );
        write_focus(client, &emissions, &mut buf)?;
        for frame in &buf {
            client.send_checked(frame).await?;
        }
    }
    Ok(start.elapsed())
}

// Keep `max_inflight` frames on the wire; drain responses behind the send cursor.
async fn run_streaming(
    client: &Client,
    geometry: &Geometry,
    targets: &[Point3<f32>],
    wavelength: Length,
    max_inflight: usize,
) -> Result<Duration> {
    let mut emissions = geometry.pattern_buffer();
    let mut buf = Frames::default();
    let mut pending: VecDeque<ResponseFuture> = VecDeque::with_capacity(max_inflight);

    let start = Instant::now();
    for &target in targets {
        autd3_rs_pattern::focus(
            geometry,
            target,
            wavelength,
            &autd3_rs_pattern::FocusOption::default(),
            &mut emissions,
        );
        write_focus(client, &emissions, &mut buf)?;
        for frame in &buf {
            if pending.len() >= max_inflight {
                pending.pop_front().expect("non-empty").await?.check()?;
            }
            pending.push_back(client.send(frame).await?);
        }
    }
    while let Some(fut) = pending.pop_front() {
        fut.await?.check()?;
    }
    Ok(start.elapsed())
}

async fn configure(client: &Client, geometry: &Geometry) -> Result<()> {
    let mut emissions = geometry.pattern_buffer();
    autd3_rs_pattern::null(&mut emissions);
    let mut builder = client.datagram_builder();
    builder
        .push(WritePatternBuffer {
            bank: PatternBank::B0,
            index: 0,
            emissions: &emissions,
        })
        .push(ConfigPattern {
            bank: PatternBank::B0,
            config: SamplingConfig::FREQ_4K,
            size: 1,
            loop_behavior: LoopBehavior::Infinite,
        });
    let frames = builder.build()?;
    for frame in &frames {
        client.send_checked(frame).await?;
    }
    Ok(())
}

fn write_focus(
    client: &Client,
    emissions: &[[Emission; NUM_TRANSDUCERS]],
    buf: &mut Frames,
) -> Result<()> {
    let mut builder = client.datagram_builder();
    builder.push(WritePatternBuffer {
        bank: PatternBank::B0,
        index: 0,
        emissions,
    });
    builder.build_into(buf)?;
    Ok(())
}

fn report(label: &str, elapsed: Duration) {
    let rate = (TOTAL_POINTS as f64) / elapsed.as_secs_f64();
    println!("{label}: {TOTAL_POINTS} updates in {elapsed:.2?} ({rate:.0} updates/s)");
}
